writer_test.go

```diff
@@ -32,17 +32,40 @@ func TestWriteAll(t *testing.T) {
 			t.Fatal(err)
 		}
 
+		err = w.Flush()
+		if err != nil {
+			t.Fatal(err)
+		}
+
 		if b.String() != cases[i].output {
 			t.Fatalf("case %v expected output %#v, got %#v", i, cases[i].output, b.String())
 		}
 	}
 }
 
 func TestWriteAllErrorPassthrough(t *testing.T) {
-	w := NewWriter(writerStub{})
+	w := NewWriter(&writerStub{})
 	err := w.WriteAll([][]string{{""}})
 
+	if err == nil {
+		err = w.Flush()
+	}
+
 	if err != stubError {
 		t.Fatalf("expected %v got %v", stubError, err)
 	}
 }
+
+func TestWriteAllBufferErrorPassthrough(t *testing.T) {
+	for i := 0; i < 5; i++ {
+		w := NewWriter(&writerStub{})
+
+		// Replace the internal bufio buffer with a stub that fails after i "written" bytes
+		w.w = &writerStub{errorCountdown: i}
```

Comment:
Now I see the interface is for testing. How about testing with something like this, and avoid the interface in the `Writer` struct:

```go
type writeFailer() struct {}

func (w *writeFailer) Write(p []byte) (n int, err error) {
  return -1, errors.New("failing as a test case")
}
```

Reply:
Which type would you define Writer.w as in this case?

Reply:
Ha, never mind. I now notice that the WriteFailer is wrapped in your last line.

That just leave the issue that I cannot create a `bufio.Writer` with a buffer length of 0, which is the only way to include this line in a test: https://github.com/cego/go-csv/blob/buffered-write/writer.go#L37

Reply:
Well... I was wrong. It has now been simplified.